pixeltable/exec/exec_context.py

```diff
@@ -1,42 +1,210 @@
+        self.elapsed_time_task_id = None
+        self.progress_reporters = {}
+
+        self.batch_size = batch_size
         self.profile = exprs.ExecProfile(row_builder)
-        # num_rows is used to compute the total number of computed cells used for the progress bar
-        self.num_rows = None
-        self.conn = None  # if present, use this to execute SQL queries
+        self.conn = None
         self.pk_clause = pk_clause
-        self.num_computed_exprs = num_computed_exprs
         self.ignore_errors = ignore_errors
         self.random_seed = random.randint(0, 1 << 63)
+
+    def add_progress_reporter(self, desc: str, unit: str) -> 'ExecContext.ProgressReporter':
+        """Records new ProgressReporter for the given desc/unit, or returns the existing one."""
+        assert self.progress is not None
+        key = f'{desc}_{unit}'
+        if key in self.progress_reporters:
+            return self.progress_reporters[key]
+        reporter = self.ProgressReporter(self, desc, unit)
+        self.progress_reporters[key] = reporter
+        return reporter
+
+    def start_progress(self) -> None:
+        """Create Progress object and start the timer. Idempotent."""
+        if not self.show_progress or self.progress is not None:
+            return
+
+        self.progress = Progress(
+            TextColumn('[progress.description]{task.description}', table_column=Column(min_width=40)),
+            AdaptiveNumericColumn(),
+            TextColumn('[progress.completed] {task.fields[unit]}', justify='left'),
+            # ' ',
+            # TextColumn('[progress.percentage]{task.fields[rate]}[/progress.percentage]', justify='right'),
+            transient=True,  # remove after stop()
+        )
+        self.progress.start()
+        self.progress_start = time.monotonic()
+
+    def stop_progress(self) -> None:
+        """Stop the timer and print the final progress report. Idempotent."""
+        if not self.show_progress or self.progress is None:
+            return
+
+        self.progress.stop()
+        # for some reason, the progress bar is not cleared automatically in jupyter
+        if getattr(builtins, '__IPYTHON__', False):
+            from IPython.display import clear_output
+
+            clear_output(wait=True)
+
+        # report the total throughput of the last stage
+        last_reporter = next(reversed(self.progress_reporters.values()))
+        last_total = last_reporter.total
+        elapsed = time.monotonic() - self.progress_start
+        self.progress.console.print(
+            f'{_print_number(last_total)} {last_reporter.unit} in {elapsed:.2f}s '
+            f'({last_total / elapsed:.2f}{last_reporter.unit}/s)'
+        )
+
+
+def _print_number(val: float | int) -> str:
+    if isinstance(val, int):
+        return str(val)
+    else:
+        assert isinstance(val, float)
+        if val < 1.0:
+            return f'{val:.3f}'
+        elif val < 10.0:
+            return f'{val:.2f}'
```

Comment:
Suggestion: `f'{val:.2f} '` (with an extra space) to align the decimal point. Likewise add extra spaces below to keep the decimal point aligned in all cases. This will avoid distracting "blips" when the order of magnitude changes, and also avoid hard-to-read tables with misaligned decimal points.

Reply:
Done.